### ISSUE TITLE
Browser mains etc

### ISSUE THREAD

**Issue by guybedford** [zenparsing/node-default-module-proposal#13]
Not directly related to Node but worth thinking about - another one of the reasons I previously backed off the `module` approach was due to interactions with other main types - for example there is a `browser`, `electron` and a `react-native` main in common use.

Ideally there should be a story for these to upgrade to modules using this approach.

Perhaps having the `default.js` analog helps here though.

`browser` does support a map, which could in theory provide a `browser: { 'index': 'browser-legacy-main.js', 'default': 'browser-module-main.js' }` map thereby providing browser support, but it would be good to ensure a clear story here for tools.

**Comment by guybedford** [comment]
Another alternative would be to have a folder called `browser-main` containing `default.js` and `index.js`, where `"browser": "browser-main"` was in the package.json file.

**Comment by zenparsing** [comment]
> Another alternative would be to have a folder called browser-main containing default.js and index.js, where "browser": "browser-main" was in the package.json file.

This seems like a pretty decent solution.

I think one of the keys for success will be keeping any interaction with `package.json` as minimal as possible, to keep things simple and to stay out of the way of tools. I think by limiting the change to only supporting a `module` key, we're doing pretty well. That leaves lots of room for tools to come up with their own solutions.

Another option might be for tools to use "module" as a prefix for other keys.

```json
{
  "module": "src/node-entry-point.js",
  "module:browser": "src/browser-entry-point.js"
}
```

To summarize the options so far:

- Point alternative entry points to a folder with "index" and "default"
- Extend alternative entry point values to be a dictionary
- Use `module:<x>` to refer to alternative module entry points

**Comment by guybedford** [comment]
> I think one of the keys for success will be keeping any interaction with package.json as minimal as possible, to keep things simple and to stay out of the way of tools. I think by limiting the change to only supporting a module key, we're doing pretty well. That leaves lots of room for tools to come up with their own solutions.

Completely agree.

All the options sounds good, although the combination types of `module:<x>` are something that specifically worries me, and I wouldn't want to advocate at all or even suggest actually.

**Comment by zenparsing** [comment]
It seems we're satisfied with the story here.

**Comment by guybedford** [comment]
I've put together a rough spec idea here at https://gist.github.com/guybedford/dfa406c61537521cc9c4dc495b37f637 - would be interested to hear thoughts.

One reason I'm sharing this here is that removing the `default.js` directory index lookup has the problem of causing issues for the dual mode submodule cases. But the proposal there by defining a `submodules` in the package.json may provide a way of supporting this as its a new property that would only work for es module resolution.
